onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.cc

```diff
@@ -0,0 +1,179 @@
+  Status status = Status::OK();
+
+  do {
+    // TODO: for now this just needs to support ONNX ops. If we ever had a transformer that was going to
+    // target non-ONNX ops we'd need to rework a few things to include the op domain in the matches
+    if (node.Domain() != kOnnxDomain) {
+      break;
+    }
+
+    auto op_rule = op_type_to_selector_and_action_.find(node.OpType());
+    if (op_rule == op_type_to_selector_and_action_.cend()) {
+      break;
+    }
+
+    const auto& selector_and_actions = *op_rule->second;
+
+    // check the supported versions if specified
+    const auto& versions = selector_and_actions.ops_and_versions.find(node.OpType())->second;
+    if (!versions.empty()) {
+      if (std::find(versions.cbegin(), versions.cend(), node.SinceVersion()) == versions.cend()) {
+        break;
+      }
+    }
+
+    std::unique_ptr<NodesToOptimize> node_group;
+    if (!selector_and_actions.selector->Select(graph, node, node_group)) {
+      break;
+    }
+
+    LOGS(logger, VERBOSE) << "Matched " << node.OpType();
+
+    if (save_) {
+      // TODO: save to Graph using transformer and action name so the node groups and actions are scoped to a
+      // specific transformer.
+      // e.g. map<transformer name, map<action name, vector<NodesToOptimizeIndexes>>>
+      ORT_NOT_IMPLEMENTED("TODO: Save the selected nodes into the Graph.");
+    } else {
+      status = selector_and_actions.action->Run(graph, *node_group);
+      if (!status.IsOK()) {
+        break;
+      }
+
+      modified = true;
+    }
+  } while (false);
+
+  return status;
+}
+#else
+void SelectorsAndActions::RegisterAction(const std::string& name,
+                                         std::unique_ptr<Action> action) {
+  ORT_ENFORCE(actions_map_.find(name) == actions_map_.cend(), "Existing registration with name ", name);
+
+  ORT_IGNORE_RETURN_VALUE(actions_map_.emplace(name, std::move(action)));
+}
+
+// TODO: The implementation here is purely an example to give an idea of how it might be done.
+//
+// The optimization info would be most conveniently stored in the Graph instance under the transformer name
+// as that makes de/serialization to ORT format simple (done as part of Graph de/serialization)
+// as well as handling subgraphs (values are stored in the current graph, be that the main graph or the subgraph).
+struct ActionReplay {
+  const std::string action_name;
+  std::vector<NodesToOptimizeIndexes> node_groups;
+};
+
+Status SelectorActionTransformer::ApplySaved(Graph& graph, bool& modified, const logging::Logger& /*logger*/) const {
+  auto fake_get_saved_actions = [](const std::string& /*transfomer_name*/) {
+    return std::vector<ActionReplay>();
+  };
+
+  // retrieve any actions saved by this transformer in the Graph
+  // TODO - setup infra for this to come from the Graph instance. could also have a 'clear all' method to free
+  // all the saved actions once we're done replaying them.
```

Comment:
How will SelectorActionTransformer be called in ort mobile?

Reply:
It will be similar to the current transformers but not quite the same.

If there are SelectorActionTransformer (SAT) instances registered for custom EPs (NNAPI and CoreML) those have to run first. 'run' most likely being a call to the Apply method with the Graph instance if we make the info on nodes to optimize for each transformer accessible via the Graph. 

Following that we need to run the GraphPartitioner so the custom EPs can do their node fusion. 

After that, any remaining SAT instances can run in priority order (e.g. CPU QDQ would run before an NHWC SAT transformer if we added that). Same order as they would run in the full build. 

Each SAT instance will only be run once though (vs. we iterate through the transformers in a level as long as something changes the graph). As they have a fixed set of nodes they're looking for we don't expect some other SAT transformer to add a node that a different SAT transformer is looking for. In theory we could support that but you'd need a deterministic NodeIndex for a node being added by an SAT transformer, and AFAIK there's no scenario where there's overlap between the changes each potential SAT transformer (level 2 or higher only) would make.